### DISCUSSION TITLE
is_instance for 3rd party packages

### DISCUSSION THREAD

**Discussion by pravic** [PyO3/pyo3#2195] [Questions]
refs #1914.

There's [`PyAny::is_instance::<T>()`](https://docs.rs/pyo3/0.15.1/pyo3/prelude/struct.PyAny.html#method.is_instance), however it's only for statically known types. 

There's also [`PyType::is_instance(&obj)`](https://docs.rs/pyo3/0.15.1/pyo3/types/struct.PyType.html#method.is_instance) but it needs an instance of a checked type.

And, there's [`pyo3::import_exception!`](https://docs.rs/pyo3/0.15.1/pyo3/macro.import_exception.html) macro which _looks like_ it can statically import a type even though it's for exceptions only.

What's the best approach for something like the following?

```py
from mypackage import MyType

def is_my_type(obj: any) -> bool:
  return isinstance(obj, MyType)

```

By looking at `import_exception!` I've come up with this:

```rust
    fn is_my_type(obj: &PyAny) -> PyResult<bool> {
      let py = obj.py();
      let module = py.import("mypackage")?;
      let klass = module.getattr("MyType")?;
      let ty: &pyo3::types::PyType = klass.extract()?;
      ty.is_instance(obj)
    }
```

Is there anything better? Or it's idiomatic enough?

**Comment by birkenfeld** [comment]
I think that's as good as it gets. Of course, once you have a reference to the type object you can cache it somewhere when convenient.
